lib/urlapi.c

```diff
@@ -596,16 +599,37 @@ static CURLUcode hostname_check(char *hostname, unsigned int flags)
 
   if(hostname[0] == '[') {
     hostname++;
-    l = "0123456789abcdefABCDEF::.%";
+    l = "0123456789abcdefABCDEF::.";
     hlen -= 2;
   }
 
   if(l) {
     /* only valid letters are ok */
     len = strspn(hostname, l);
-    if(hlen != len)
-      /* hostname with bad content */
-      return CURLUE_MALFORMED_INPUT;
+    if(hlen != len) {
+      /* this could now be '%[zone id]' */
+      char scopeid[16];
+      if(hostname[len] == '%') {
+        int i = 0;
+        char *h = &hostname[len + 1];
+        /* pass '25' if present */
+        if(!strncmp(h, "25", 2))
```

Comment:
According to [RFC 6874](https://tools.ietf.org/html/rfc6874#page-4) `IPv6addrz = IPv6address "%25" ZoneID` so I guess just % is being allowed for backwards compatibility? In Windows 7 the zone identifiers I see are numeric, like for example I have a 20 - 28 on my computer for virtual adapters, so for example `[fe80::20c:29ff:fe9c:409b%25]` and that could be allowed with 
~~~diff
diff --git a/lib/urlapi.c b/lib/urlapi.c
index c666d55..557edd2 100644
--- a/lib/urlapi.c
+++ b/lib/urlapi.c
@@ -612,12 +612,12 @@ static CURLUcode hostname_check(struct Curl_URL *u,
       if(hostname[len] == '%') {
         int i = 0;
         char *h = &hostname[len + 1];
-        /* pass '25' if present */
-        if(!strncmp(h, "25", 2))
+        /* skip '25' if present and is a url encoded percent sign */
+        if(!strncmp(h, "25", 2) && h[2] && (h[2] != ']'))
           h += 2;
         while(*h && (*h != ']') && (i < 15))
           scopeid[i++] = *h++;
-        if(']' != *h)
+        if((i <= 1) || (']' != *h))
           return CURLUE_MALFORMED_INPUT;
         scopeid[i] = 0;
         u->scopeid = strdup(scopeid);
~~~

also shouldn't existing u->scopeid be freed when u->host is going to be set (I don't see that here but I assume it happens somewhere)

Reply:
> just % is being allowed for backwards compatibility?

Yes, because that's what we used to support and that's something people keep occasionally throw on curl. As #3817 shows.

> like for example I have a 20 - 28 on my computer

Oh. Thanks, that's indeed something to take into account. I'll get your patch in and add a test for that case.

> shouldn't existing u->scopeid be freed when u->host is going to be set

Tricky question. We don't do that association for other parts so I'm not sure we should for the scope id either, but then it isn't exactly like the other parts...